### PR TITLE
Implement sysrepo notification daemon (sysrepo-notifd)

### PR DESCRIPTION
**Pull request by Roytak**
## Summary

- Implement `sysrepo-notifd`, a standalone notification daemon that handles `ietf-subscribed-notifications` configured subscriptions notification replay, storage, and UDP relay
- Store and expose the replay-start timestamp (when replay was first enabled) to support `replay-log-creation-time` from `ietf-subscribed-notifications`
- Add test suite for the notification daemon (`test_notifd`)

## API Changes

**`sr_get_module_replay_support()`** - unchanged, kept original signature:
```c
int sr_get_module_replay_support(sr_conn_ctx_t *conn, const char *module_name,
        struct timespec *earliest_notif, int *enabled);
```

**`sr_get_module_replay_start()`** - new function:
```c
int sr_get_module_replay_start(sr_conn_ctx_t *conn, const char *module_name,
        const struct timespec *after, struct timespec *earliest_notif, struct timespec *replay_start);
```
- Parameter `after`: optional filter to get the earliest stored notification at or after this time
- Parameter `earliest_notif`: optional timestamp of the earliest stored notification (at or after `after`)
- Parameter `replay_start`: returns when replay was enabled, zeroed if replay is currently disabled

`Notification Plugin API`:

- **`srntf_earliest_get`** callback signature changed: added `const struct timespec *after` parameter
- New callback **`srntf_replay_start_get`**: returns the timestamp when replay was enabled for a module

### CODE REVIEW

File: src/executables/sysrepo-notifd/sysrepo-notifd-config.c [hunk 1]
```diff
@@ -0,0 +1,1821 @@
...
+        *encoding = NOTIF_ENCODING_JSON;
+    } else if (!strcmp(value, "encode-cbor")) {
+        SRNTF_LOG_ERR("CBOR encoding is not yet supported.");
+        return SR_ERR_UNSUPPORTED;
+    } else {
+        SRNTF_LOG_ERR("Unsupported encoding \"%s\".", value);
+        return SR_ERR_UNSUPPORTED;
+    }
+
+    return SR_ERR_OK;
```
Comment (michalvasko) [inline src/executables/sysrepo-notifd/sysrepo-notifd-config.c]: Unreachable.

File: src/executables/sysrepo-notifd/sysrepo-notifd-runtime.c [hunk 2]
```diff
@@ -0,0 +1,757 @@
...
+
+static int
+subscription_terminated_notif_new(const struct ly_ctx *ly_ctx, notif_sub_t *sub, const char *reason,
+        struct lyd_node **notif)
+{
+    return subscription_state_change_notif_new(ly_ctx, sub,
+            "/ietf-subscribed-notifications:subscription-terminated", 0, reason, notif);
+}
+
+static int __attribute__((unused))
```
Comment (michalvasko) [inline src/executables/sysrepo-notifd/sysrepo-notifd-runtime.c]: Unused code is usually wrapped in `#if 0`.

File: src/executables/sysrepo-notifd/sysrepo-notifd-runtime.c [hunk 3]
```diff
@@ -0,0 +1,757 @@
...
+        SRNTF_LOG_ERR("Cannot send notification \"%s\" to receiver \"%s\" before sending subscription-started.",
+                notif_path, receiver->name);
+        rc = SR_ERR_OPERATION_FAILED;
+        goto cleanup;
+    }
+
+    if (!timestamp) {
+        /* get the current time */
+        now = time(NULL);
+        ts.tv_sec = now;
```
Comment (michalvasko) [inline src/executables/sysrepo-notifd/sysrepo-notifd-runtime.c]: Use the full `timespec` precision using functions such as `clock_gettime()`.

File: src/executables/sysrepo-notifd/notifd_runtime.c [hunk 4]
```diff
@@ -0,0 +1,757 @@
...
+    }
+
+    if ((receiver->state != NOTIF_RECV_STATE_ACTIVE) && !is_sub_started) {
+        SRNTF_LOG_ERR("Cannot send notification \"%s\" to receiver \"%s\" before sending subscription-started.",
+                notif_path, receiver->name);
+        rc = SR_ERR_OPERATION_FAILED;
+        goto cleanup;
+    }
+
+    if (!timestamp) {
```
Comment (michalvasko) [inline src/executables/sysrepo-notifd/notifd_runtime.c:463]: Can this situation even occur?

File: src/executables/sysrepo-notifd/sysrepo-notifd.c [hunk 5]
```diff
@@ -0,0 +1,1500 @@
...
+    if (sr_rpc_subscribe_tree(notifd_ctx.sr_sess,
+            "/ietf-subscribed-notifications:subscriptions/subscription/receivers/receiver/reset",
+            receiver_reset_rpc_cb, &notifd_ctx, 0, 0, &sr_subscr)) {
+        SRNTF_LOG_ERR("Failed to subscribe for receiver reset action.");
+        rc = EXIT_FAILURE;
+        goto cleanup;
+    }
+
+    /* update pid file */
+    if (pidfile && (write_pidfile(pidfd) < 0)) {
```
Comment (michalvasko) [inline src/executables/sysrepo-notifd/sysrepo-notifd.c]: It is better to create the PID file and write into it at the same time, before notifying `systemd`. That way, once the PID file exists, any application can check the app has fully started and does not need to read the file.

File: src/ly_wrap.c [hunk 6]
```diff
@@ -306,6 +306,16 @@ sr_ly_ctx_new(struct ly_ctx **ly_ctx)
     return err_info;
 }
 
+void
+sr_ly_ctx_destroy(struct ly_ctx *ly_ctx)
+{
+    sr_ly_log_setup();
+
+    ly_ctx_destroy(ly_ctx);
```
Comment (michalvasko) [inline src/ly_wrap.c:314]: Any warnings or errors generated by this function should be printed, check it is so.

File: src/sysrepo.c [hunk 7]
```diff
@@ -2269,7 +2269,8 @@ sr_set_module_replay_support(sr_conn_ctx_t *conn, const char *module_name, int e
 }
 
 API int
-sr_get_module_replay_support(sr_conn_ctx_t *conn, const char *module_name, struct timespec *earliest_notif, int *enabled)
+sr_get_module_replay_support(sr_conn_ctx_t *conn, const char *module_name, const struct timespec *after,
```
Comment (michalvasko) [inline src/sysrepo.c]: Do not change the API. You can deprecate this function and replace it with a new one or simply add a new one for the `replay_start` parameter.

File: tests/test_notifd.c [hunk 8]
```diff
@@ -0,0 +1,3216 @@
...
+ * @param[in] message_id Message ID.
+ * @param[in] media_type Media type.
+ * @return Pending message slot or NULL on error.
+ */
+static pending_message_t *
+find_or_create_pending(uint32_t publisher_id, uint32_t message_id, uint8_t media_type)
+{
+    pending_message_t *oldest = NULL;
+    time_t oldest_time = 0;
+    time_t now = time(NULL);
```
Comment (michalvasko) [inline tests/test_notifd.c:225]: Check second-precision is really okay and cannot cause any corner-case failures.

**Review by michalvasko — CHANGES_REQUESTED** [review]

File: src/executables/sysrepo-notifd/sysrepo-notifd-config.c (hunk 1 above)
Comment (Roytak) [reply src/executables/sysrepo-notifd/sysrepo-notifd-config.c]: Was reachable, but changed the code to be more readable.

File: src/executables/sysrepo-notifd/main.c [hunk 9]
```diff
@@ -0,0 +1,1488 @@
...
+
+    /* reset the receiver state to connecting, it will attempt to reconnect once there is a notif to send */
+    recv->state = NOTIF_RECV_STATE_CONNECTING;
+
+    /* reset reconnect backoff */
+    recv->reconnect_attempts = 0;
+    memset(&recv->last_reconnect_attempt, 0, sizeof(recv->last_reconnect_attempt));
+
+    /* create the output - current time as the reset time */
+    time_now = time(NULL);
```
Comment (michalvasko) [inline src/executables/sysrepo-notifd/main.c]: Use `timespec` precision.

**Review by michalvasko — CHANGES_REQUESTED** [review]

**Review by michalvasko — APPROVED** [review]
